### PR TITLE
Update the error to expected

### LINKED ISSUE #8470: SSL_get_error return value was not expect in some case
Version: OpenSSL1.1.1
Scenario: 
      The application call some other OpenSSL API and did not call ERR_clear_error() in every place. and SSL_read return failed in case of socket non-block error. and the expected error currently is SSL_ERROR_WANT_READ. but the actual error was SSL_ERROR_SSL, as the error stack have keeped some other error which have pushed before SSL_read. 
Adapter:
...
Some other OpenSSL API. //error stack did not clear with ERR_clear_error().
...
k = SSL_read(con, sbuf, 1024 /* BUFSIZZ */ );
switch (SSL_get_error(con, k)) {
...
case SSL_ERROR_WANT_WRITE:
...
case SSL_ERROR_WANT_READ:
...
case SSL_ERROR_SSL: 
...
}

Analysis:
The error stack did not clear in the beginning of SSL_read and SSL_get_error will try to get the error with ERR_peek_error firstly. state_machine have clear the error, so SSL_dohandshake  will not report such error.

int SSL_get_error(const SSL *s, int i){
...
    if ((l = ERR_peek_error()) != 0) {
        if (ERR_GET_LIB(l) == ERR_LIB_SYS)
            return SSL_ERROR_SYSCALL;
        else
            return SSL_ERROR_SSL;
    }
    if (SSL_want_read(s)) {
        bio = SSL_get_rbio(s);
        if (BIO_should_read(bio))
            return SSL_ERROR_WANT_READ;
        else if (BIO_should_write(bio))
            return SSL_ERROR_WANT_WRITE;
...
}

static int state_machine(SSL *s, int server){
...
    ERR_clear_error();
    clear_sys_error();
...
}

Suggestion:
1 Clear the error stack at the beginning of SSL_read/SSL_write/SSL_shutdown.
2 move the SSL_want_read logic before the ERR_peek_error,so that the SSL_ERROR_WANT_READ/ SSL_ERROR_WANT_WRITE have the high priority and will reported Correctly when the application missed to clear some error before SSL_read.
int SSL_get_error(const SSL *s, int i){
...
    if ((l = ERR_peek_error()) != 0) {
        if (ERR_GET_LIB(l) == ERR_LIB_SYS)
            return SSL_ERROR_SYSCALL;
        else
            return SSL_ERROR_SSL;
    }
    if (SSL_want_read(s)) {
        bio = SSL_get_rbio(s);
        if (BIO_should_read(bio))
            return SSL_ERROR_WANT_READ;
        else if (BIO_should_write(bio))
            return SSL_ERROR_WANT_WRITE;
...
}

### PR DESCRIPTION
**Pull request by KanPlus**
Fixes #8470

##### Checklist

- [x] tests are added or updated

### CODE REVIEW

**Comment by t8m** [comment]
The CI is relevant.

**Comment by KanPlus** [comment]
@mattcaswell 
As  the base of #17169 update so much, This PR was raised.
1 "s->statem.state" with MSG_FLOW_ERROR could not cover the case that error reported from this peer, As ERR_raise called without set the state.
2 BIO status will kept and impact the new error from ERR_raise.

File: ssl/ssl_lib.c [hunk 1]
```diff
@@ -3812,19 +3817,16 @@ int SSL_get_error(const SSL *s, int i)
...
-    /*
-     * Make things return SSL_ERROR_SYSCALL when doing SSL_do_handshake etc,
-     * where we do encode the error
-     */
-    if ((l = ERR_peek_error()) != 0) {
-        if (ERR_GET_LIB(l) == ERR_LIB_SYS)
-            return SSL_ERROR_SYSCALL;
-        else
-            return SSL_ERROR_SSL;
+    if((s == NULL) || ossl_statem_in_error(s)) {
```
Comment (FdaSilvaYY) [inline ssl/ssl_lib.c:3820]: style nit : should be : 
`    if (s == NULL || ossl_statem_in_error(s)) {`

**Review by t8m — CHANGES_REQUESTED** [review]

File: test/ssl-tests/17-renegotiate.cnf.in [hunk 2]
```diff
@@ -228,7 +228,7 @@ our @tests_tls1_2 = (
             "Method" => "TLS",
             "HandshakeMode" => "RenegotiateServer",
             "ResumptionExpected" => "No",
-            "ExpectedResult" => "ServerFail"
+            "ExpectedResult" => "InternalError"
```
[the remark this answers is not in the document]
Comment (KanPlus) [reply test/ssl-tests/17-renegotiate.cnf.in:231]: The ssl_test framework did not handle the case, I will check How to update,thanks.

**Comment by t8m** [comment]
From high level POV it looks OK. However it is unclear to me whether this is a bug fix or feature. Also I am not sure enough this is a right fix to provide approval.

Perhaps @mattcaswell @kaduk could review?

File: ssl/ssl_lib.c [hunk 3]
```diff
@@ -3884,6 +3883,17 @@ int SSL_get_error(const SSL *s, int i)
...
+    /*
+     * Make things return SSL_ERROR_SYSCALL when doing SSL_do_handshake etc,
+     * where we do encode the error
+     */
+    if ((l = ERR_peek_error()) != 0) {
+        if (ERR_GET_LIB(l) == ERR_LIB_SYS)
+            return SSL_ERROR_SYSCALL;
+        else
+            return SSL_ERROR_SSL;
+    }
```
Comment (mattcaswell) [inline ssl/ssl_lib.c:3895]: Why is this necessary? Any error set by libssl should have been caught by `ossl_statem_in_error` above

File: test/helpers/handshake.c [hunk 4]
```diff
@@ -1609,6 +1609,18 @@ static HANDSHAKE_RESULT *do_handshake_internal(
...
+                     * The server Failed and result client non-block.
+                     */
+                    ret->result = SSL_TEST_SERVER_FAIL;
+                    goto err;
+                } else if (!client_turn && client.status == PEER_ERROR) {
+                    /*
+                     * The client Failed and result server non-block.
+                     */
+                    ret->result = SSL_TEST_CLIENT_FAIL;
+                    goto err;
```
Comment (mattcaswell) [inline test/helpers/handshake.c:1623]: Why is this change necessary? It suggests a change of behaviour which is undesirable.

**Review by mattcaswell — CHANGES_REQUESTED** [review]
I broadly support the idea of using `ossl_statem_in_error()` to decide whether to return `SSL_ERROR_SSL` from `SSL_get_error()`. This is a much more robust approach than examining the error stack. However I consider this to be a very significant change in approach and could imply a change of behaviour that I am not entirely comfortable with applying in a minor release. I suspect it will have lots of unforeseen consequences and corner cases. This is probably a more appropriate changes for a major release. This is one reason why I haven't already made a similar change myself - I've been thinking about doing this for some while.

File: ssl/ssl_lib.c (hunk 3 above)
Comment (KanPlus) [reply ssl/ssl_lib.c:3895]: ossl_statem_in_error could not handle the error from application protocol,this kept for compatibility.

File: test/helpers/handshake.c (hunk 4 above)
Comment (KanPlus) [reply test/helpers/handshake.c:1623]: It  behavior should be compatible and still kept with the update expected error, **The test case:** this side return failed and did not send message . and the  peer detect the non-block and then check with the prior status, this case, the root error should from this side.
The prior error code: this side:ERROR, the peer:ERROR
The current error code: this side:ERROR, the peer:RETRY

Comment (KanPlus) [reply test/helpers/handshake.c:1623]: ossl_statem_in_error could not handle the error from application protocol,this kept for compatibility.And this will remove once all other case update the statem error.

**Comment by KanPlus** [comment]
> I broadly support the idea of using `ossl_statem_in_error()` to decide whether to return `SSL_ERROR_SSL` from `SSL_get_error()`. This is a much more robust approach than examining the error stack. However I consider this to be a very significant change in approach and could imply a change of behaviour that I am not entirely comfortable with applying in a minor release. I suspect it will have lots of unforeseen consequences and corner cases. This is probably a more appropriate changes for a major release. This is one reason why I haven't already made a similar change myself - I've been thinking about doing this for some while.

Thanks for your review, this only update the error to expected and did not add any new feature from the user side actually.

**Comment by nhorman** [comment]
Closing for inactivity
